Functionality to detect Win32 apps which are installed, deleted or renamed while PowerToys is running

alekhyareddy28 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
When a Win32 app is installed, deleted or renamed, PowerToys Run updates itself accordingly.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #3553
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
The changes made in this PR are as follows -
* The `win32Storage` list has been modified to use a class, called `Win32ProgramRepository` which implements `ListRespository` and `IProgramRepository` just like in packaged applications. With this change all the items are stored in a concurrent dictionary and adding, deleting, indexing all programs has become easier. As concurrent dictionary is thread safe, we do not need locks while setting/deleting items.
* When a Win32 application is added, there can be a chance of duplicates. To avoid this situation, the `GetHashCode()` function of Win32 has been overridden so that the custom Win32 GetHashCode() function is used. This ensures that no duplicates are present in the list even when we actively add or delete apps.
* The `Win32ProgramRepository` and `Win32ProgramRepositoryHelper` classes have been added to watch certain locations and trigger `Created`, `Deleted` or `Renamed` events whenever such an operation is performed on a file. To prevent these events from being triggered for non app type files, we add `Filters`. These filters raise events only for those particular type of applications. [Filters](https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher.filters?view=netcore-3.1)
* The `FileSystemWatchers` filter out exe, lnk, appred-ms and url type files and call the `GetAppFromPath(path)` function which has been added to the Win32 class. This function returns the corresponding Win32 item which can be added to the list. 
* Wrappers and interfaces have been added to `FileSystemWatcher`, `System.IO.File` and `FileVersionInfo` classes so that they can be mocked and tested.
* `ShellLinkHelper` is implementing the interface `IShellLinkHelper` so that the Win32 shortcut programs (lnk) can be tested.

* For shortcut apps, `GetAppFromPath` doesn't give us the expected FullPath for the deleted and renamed application and therefore it must be treated in a separate way because the FullPath is set using the ShellLinkHelper and if the file does not exist we cannot extract the **exe FullPath** from the lnkFile.
* For `Deleted` apps since we have no more information to go on, we iterate through all the programs and remove the one which has the same lnkResolved path.
* For `Renamed` apps, we have more information about the existing app, so we construct what could have been the old app so that we can remove it from the list. This is done so that we have an app which has the same name, exe name and full path as the app which was added so that the hashcodes are the same.
<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
* Manually validated it.
* Added tests -
1.  To ensure that the we will not be introducing duplicates.
2. For each event {Created, Renamed, Deleted} and app type {exe, lnk, appred-ms, url} a test has been added to validate that the FileSystemWatcher work as expected.
* All the other program plugin tests pass as well.
![image](https://user-images.githubusercontent.com/28739210/87326150-368e3400-c4e7-11ea-84d5-3f7f14ad7be2.png)

crutkas commented:
@bao-qian and @jjw24 this PR may be of interest to you.

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/IShellLinkHelper.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public interface IShellLinkHelper
+    {
+        string retrieveTargetPath(string path);
```
nit: First character of methods should be capitalized

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32.cs:
```diff
@@ -36,6 +37,10 @@ public class Win32 : IProgram
         public string Location => ParentDirectory;
         public uint AppType { get; set; }
+        // Wrappers for File Operations
+        public static IFileVersionInfoWrapper fileVersionInfoWrapper = new FileVersionInfoWrapper();
```
nit: Inconsistent variable naming -- underscore vs no underscore.

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32.cs:
```diff
@@ -439,16 +444,16 @@ private static Win32 ExeProgram(string path)
 
                 return program;
             }
-            catch (Exception e) when (e is SecurityException || e is UnauthorizedAccessException)
```
@alekhyareddy28 Removing this could hide exceptions that we are aren't expecting.  Why did we need to remove this.

alekhyareddy28 replied:
Thanks for pointing it out Ryan. I had removed it before I added a wrapper for System Operations and I had forgotten to add that back. Reverted it to how it was.

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Storage/Win32ProgramRepository.cs:
```diff
@@ -0,0 +1,202 @@
+        private readonly string lnkExtension = ".lnk";
+        private readonly string urlExtension = ".url";
+
+        public Win32ProgramRepository(List<IFileSystemWatcherWrapper> fileSystemWatcherHelpers, IStorage<IList<Win32>> storage, Settings settings, string[] pathsToWatch)
```
nit: this could be slightly more flexible if it takes `IList<IFileSystemWatcherWrapper>`

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Storage/Win32ProgramRepositoryHelper.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace Microsoft.Plugin.Program.Storage
+{
+    internal class Win32ProgramRepositoryHelper
```
nit:  I'm not a big fan of the term 'helper', what about Win32ProgramFileSystemWatchers?

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Storage/Win32ProgramRepositoryHelper.cs:
```diff
@@ -0,0 +1,43 @@
+            _fileSystemWatchers = new List<FileSystemWatcherWrapper>();
+            for (int index = 0; index < _pathsToWatch.Length; index++)
+            {
+                _fileSystemWatchers.Add(new FileSystemWatcherWrapper());
```
I think we need to dispose these.

alekhyareddy28 replied:
Completely agree on this. Shall update it accordingly.

somil55 commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Main.cs:
```diff
@@ -22,36 +21,39 @@ namespace Microsoft.Plugin.Program
         private readonly PluginJsonStorage<Settings> _settingsStorage;
         private bool _disposed = false;
         private PackageRepository _packageRepository = new PackageRepository(new PackageCatalogWrapper(), new BinaryStorage<IList<UWP.Application>>("UWP"));
+        private Win32ProgramRepositoryHelper _win32ProgramRepositoryHelper;
```
[nit] This could be `static`.

somil55 commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Main.cs:
```diff
@@ -22,36 +21,39 @@ namespace Microsoft.Plugin.Program
         private bool _disposed = false;
         private PackageRepository _packageRepository = new PackageRepository(new PackageCatalogWrapper(), new BinaryStorage<IList<UWP.Application>>("UWP"));
+        private Win32ProgramRepositoryHelper _win32ProgramRepositoryHelper;
+        private Win32ProgramRepository _win32ProgramRepository;
```
[nit] This could be `static`.

somil55 commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Main.cs:
```diff
@@ -64,26 +66,19 @@ public Main()
-            }
-
-            var results1 = win32.AsParallel()
+            var results1 = _win32ProgramRepository.AsParallel()
```
I think there would be a synchronization issue here in case a program is installing/uninstalling and a query is made. The earlier code was handling this.

alekhyareddy28 replied:
We are using a ConcurrentDictionary which is threadsafe. I don't think this should be an issue. @ryanbodrug-microsoft, do you think there would be an issue?

somil55 replied:
My bad. That should be fine. @ryanbodrug-microsoft can confirm.

alekhyareddy28 commented:
> I noticed the following issues while trying this PR :
> 
> 1. Results are not updated after a program is installed/uninstalled. Query needs to be manually changed to see updated results. For example :
>    a. Type in `notion` in the search box.
>    b. Uninstall `notion.exe`.
>    c.  Press hotkey to bring up the launcher. `Notion` is still present in the results.

This is not related to this PR. That is how PT Run is. The results are obtained once for each query. I don't think this is related to this PR. The user has to change the query for the results to update. This would be the same situation with even UWP applications/anything else as well. JsonRPC is used to communicate with each plugin for each query and once the results are obtained they are displayed.

> 2. If I abruptly close powerlauncher (say by closing debugging session directly), the cache for the uninstalled program is still present on the next launch. We should verify cache validity on startup.

Yea, i think it is a good suggestion. Apps which are deleted when PT Run is not running also show up because of the cache issue. I think it is a generic issue and I shall create it an issue so that we can work on it.

> 3. If I rename an internet application, it keeps all the previous version as well

I tested this out to ensure that this would not happen. Can you please share information about the app (ie. LnkResolvedPath, FullPath, exename) so that I can repro it and try to debug. 

> 4. I added a program using Epic Games which installed a `url` shortcut on desktop. But this change is not detected.

I tested it out with steam games and it worked as expected. I shall try out Epic games as well.

Regarding 1,2 I think they are not a scope of this PR. Regarding 3 and 4, I shall look into why that's happening and try to fix it.

alekhyareddy28 commented:
@somil55, I tried installing the same game on the desktop while PT Run was executing in the background and it works as expected. Would be great if you could help me figure out those two scenarios because I'm not able to repro them. Thanks!
![image](https://user-images.githubusercontent.com/28739210/87615003-77906f00-c6c6-11ea-88c9-94f497cb85ec.png)

somil55 commented:
Regarding 1, we actually do that in certain cases. For example, when theme is changed, query is refreshed to update the theme-specific icon. This was done in #4499. But I agree that 1, 2 can be done as part of separate PR. 

The app in 3 is hangouts chrome extension with the following info 
```
LnkResolvePath : c:\\users\\divyan\\appdata\\roaming\\microsoft\\windows\\start menu\\programs\\chrome apps\\google hangouts d.lnk
Fullpath : c:\\program files (x86)\\google\\chrome\\application\\chrome_proxy.exe
ExecutableName : CHROME~1.EXE
```
I am able to repro this issue each time I rename this file. Please tell me if you need any more information.

alekhyareddy28 commented:
I resolved all the comments.

A few things to point out -

1. I noticed that we were previously not disposing all the plugins properly while trying to dispose the File system watchers. We had the code for disposing events in the plugin manager but we had `var disposablePlugin = plugin as IDisposable;` instead of `var disposablePlugin = plugin.Plugin as IDisposable;` and hence, the Dispose functions weren't being called. I fixed that and ensured that all the plugins are being properly disposed of and even the File system watchers are being properly disposed.

2. Regarding the renaming of shortcut apps, I was previously just removing the extension to get the name of the app from the `OldName` Rename event handler args. However, the `Name` parameter contains the path as well with respect to the directory that we are watching. I used `Path.GetFileNameWithoutExtension(path)` and that fixes the issue.

3. Regarding the url apps showing up sometimes, it wasb' because the installation of the game was not yet complete and it was still being written to. Hence, when we tried to call the function `ReadAllLines`, it would throw an IOException. Now I have added a retry block after a set amount of time. This would not halt anything because it is on the file system watcher event handler thread.

alekhyareddy28 commented:
@somil55 I tried to fix it by adding some retry attempts. I have created an issue to look into it #5072. Can this PS be merged in as it is a huge PR and I need to be resolving merge conflicts. I shall look into that issue separately as a part of this iteration.
